File: examples/readme/basics/basics.py
"""Executable source for every Python snippet in docs/features/basics.md."""

from pathlib import Path

from nucleation import Schematic


# --8<-- [start:beacon]
from nucleation import Schematic

beacon = Schematic.create("beacon")
for x in range(-1, 2):
    for z in range(-1, 2):
        beacon.set_block(x, 0, z, "minecraft:gold_block")
beacon.set_block(0, 1, 0, "minecraft:beacon")
beacon.save_to_file("beacon.schem")
# --8<-- [end:beacon]

assert beacon.block_count() == 10
assert beacon.tight_dimensions().x == 3


# --8<-- [start:crafting-nook]
nook = Schematic.create("crafting_nook")
for x in range(5):
    for z in range(5):
        nook.set_block(x, 0, z, "minecraft:spruce_planks")

def wall_block(i, y, end_posts):
    if i == 2 and y == 2:
        return "minecraft:light_blue_stained_glass"
    if i in end_posts:
        return "minecraft:stripped_spruce_log[axis=y]"
    return "minecraft:oak_planks"

for y in (1, 2, 3):
    for x in range(5):
        nook.set_block(x, y, 0, wall_block(x, y, (0, 4)))
    for z in range(1, 5):
        nook.set_block(0, y, z, wall_block(z, y, (4,)))

nook.set_block(1, 1, 1, "minecraft:crafting_table")
nook.set_block(3, 1, 1, "minecraft:chest[facing=south]")
nook.set_block(4, 2, 1, "minecraft:wall_torch[facing=south]")
nook.set_block(1, 2, 4, "minecraft:wall_torch[facing=east]")
nook.save_to_file("crafting-nook.schem")
# --8<-- [end:crafting-nook]

assert nook.block_count() == 56


# --8<-- [start:coordinates]
build = Schematic.create("signed_coordinates")
build.set_block(-8, 64, 12, "minecraft:stone")
build.set_block(24, 80, -3, "minecraft:glass")

minimum = build.tight_bounds_min()
maximum = build.tight_bounds_max()
size = build.tight_dimensions()
print((minimum.x, minimum.y, minimum.z))  # (-8, 64, -3)
print((maximum.x, maximum.y, maximum.z))  # (24, 80, 12)
print((size.x, size.y, size.z))           # (33, 17, 16)
# --8<-- [end:coordinates]

assert (minimum.x, minimum.y, minimum.z) == (-8, 64, -3)
assert (maximum.x, maximum.y, maximum.z) == (24, 80, 12)
assert (size.x, size.y, size.z) == (33, 17, 16)


# --8<-- [start:block-states]
build = Schematic.create("inspect")
build.set_block(1, 1, 1, "minecraft:oak_log[axis=x]")
state = build.get_block(1, 1, 1)
print(state.name())                         # minecraft:oak_log
print(build.get_block_string(1, 1, 1))      # minecraft:oak_log[axis=x]

build.set_block(1, 1, 1, "minecraft:air")  # remove it
# --8<-- [end:block-states]

assert build.block_count() == 0


# --8<-- [start:contents]
contents = Schematic.create("contents")
contents.set_block(0, 0, 0, "minecraft:barrel{signal=13,item=diamond}")
contents.set_block(1, 0, 0, "minecraft:chest{items=[diamond*64,emerald*12]}")
contents.set_block(2, 0, 0, "minecraft:jukebox{record=pigstep}")
contents.set_block(3, 0, 0, "minecraft:jukebox{signal=13}")
# --8<-- [end:contents]

assert contents.block_count() == 4


# --8<-- [start:simulation]
circuit = Schematic.create("placed_by_engine")
circuit.set_block(4, 0, 0, "minecraft:redstone_block")
circuit.set_block(5, 0, 0, "minecraft:redstone_wire{simulate=true}")
print(circuit.get_block_string(5, 0, 0))
# minecraft:redstone_wire[east=side,north=none,power=15,south=none,west=side]
# --8<-- [end:simulation]

assert circuit.get_block_string(5, 0, 0) == (
    "minecraft:redstone_wire[east=side,north=none,power=15,south=none,west=side]"
)


# --8<-- [start:io]
copy = Schematic.load_from_file("beacon.schem")
copy.set_block(0, 2, 0, "minecraft:glass")
copy.save_to_file("beacon-edited.litematic")
# --8<-- [end:io]

assert Path("beacon.schem").stat().st_size > 0
assert Path("beacon-edited.litematic").stat().st_size > 0
assert copy.block_count() == 11
print("Basics Python examples: OK")

File: examples/readme/basics/basics.mjs
// --8<-- [start:beacon]
import { readFileSync, writeFileSync } from "node:fs";
import { Schematic } from "nucleation";

function bytesFromBase64(value) {
  return Uint8Array.from(Buffer.from(value, "base64"));
}

const beacon = Schematic.create("beacon");
for (let x = -1; x <= 1; x += 1) {
  for (let z = -1; z <= 1; z += 1) {
    beacon.setBlock(x, 0, z, "minecraft:gold_block");
  }
}
beacon.setBlock(0, 1, 0, "minecraft:beacon");

const beaconBytes = bytesFromBase64(beacon.toSchematicB64());
writeFileSync("beacon.schem", beaconBytes);
// --8<-- [end:beacon]

if (beacon.blockCount() !== 10) throw new Error("beacon block count changed");
if (beacon.tightDimensions().x !== 3) throw new Error("beacon width changed");


// --8<-- [start:crafting-nook]
const nook = Schematic.create("crafting_nook");
for (let x = 0; x < 5; x += 1) {
  for (let z = 0; z < 5; z += 1) {
    nook.setBlock(x, 0, z, "minecraft:spruce_planks");
  }
}

function wallBlock(i, y, endPosts) {
  if (i === 2 && y === 2) return "minecraft:light_blue_stained_glass";
  if (endPosts.includes(i)) return "minecraft:stripped_spruce_log[axis=y]";
  return "minecraft:oak_planks";
}

for (const y of [1, 2, 3]) {
  for (let x = 0; x < 5; x += 1) {
    nook.setBlock(x, y, 0, wallBlock(x, y, [0, 4]));
  }
  for (let z = 1; z < 5; z += 1) {
    nook.setBlock(0, y, z, wallBlock(z, y, [4]));
  }
}

nook.setBlock(1, 1, 1, "minecraft:crafting_table");
nook.setBlock(3, 1, 1, "minecraft:chest[facing=south]");
nook.setBlock(4, 2, 1, "minecraft:wall_torch[facing=south]");
nook.setBlock(1, 2, 4, "minecraft:wall_torch[facing=east]");
writeFileSync(
  "crafting-nook.schem",
  bytesFromBase64(nook.toSchematicB64()),
);
// --8<-- [end:crafting-nook]

if (nook.blockCount() !== 56) throw new Error("crafting nook block count changed");


// --8<-- [start:coordinates]
const build = Schematic.create("signed_coordinates");
build.setBlock(-8, 64, 12, "minecraft:stone");
build.setBlock(24, 80, -3, "minecraft:glass");

const minimum = build.tightBoundsMin();
const maximum = build.tightBoundsMax();
const size = build.tightDimensions();
console.log([minimum.x, minimum.y, minimum.z]); // [-8, 64, -3]
console.log([maximum.x, maximum.y, maximum.z]); // [24, 80, 12]
console.log([size.x, size.y, size.z]);          // [33, 17, 16]
// --8<-- [end:coordinates]

if (`${minimum.x},${minimum.y},${minimum.z}` !== "-8,64,-3") {
  throw new Error("minimum bounds changed");
}
if (`${maximum.x},${maximum.y},${maximum.z}` !== "24,80,12") {
  throw new Error("maximum bounds changed");
}
if (`${size.x},${size.y},${size.z}` !== "33,17,16") {
  throw new Error("tight dimensions changed");
}


// --8<-- [start:block-states]
const inspect = Schematic.create("inspect");
inspect.setBlock(1, 1, 1, "minecraft:oak_log[axis=x]");
console.log(inspect.getBlockName(1, 1, 1));   // minecraft:oak_log
console.log(inspect.getBlockString(1, 1, 1)); // minecraft:oak_log[axis=x]

inspect.setBlock(1, 1, 1, "minecraft:air");  // remove it
// --8<-- [end:block-states]

if (inspect.blockCount() !== 0) throw new Error("air did not remove the block");


// --8<-- [start:contents]
const contents = Schematic.create("contents");
contents.setBlock(0, 0, 0, "minecraft:barrel{signal=13,item=diamond}");
contents.setBlock(1, 0, 0, "minecraft:chest{items=[diamond*64,emerald*12]}");
contents.setBlock(2, 0, 0, "minecraft:jukebox{record=pigstep}");
contents.setBlock(3, 0, 0, "minecraft:jukebox{signal=13}");
// --8<-- [end:contents]

if (contents.blockCount() !== 4) throw new Error("content shorthand placement failed");


// --8<-- [start:simulation]
const circuit = Schematic.create("placed_by_engine");
circuit.setBlock(4, 0, 0, "minecraft:redstone_block");
circuit.setBlock(5, 0, 0, "minecraft:redstone_wire{simulate=true}");
console.log(circuit.getBlockString(5, 0, 0));
// minecraft:redstone_wire[east=side,north=none,power=15,south=none,west=side]
// --8<-- [end:simulation]

if (
  circuit.getBlockString(5, 0, 0) !==
  "minecraft:redstone_wire[east=side,north=none,power=15,south=none,west=side]"
) {
  throw new Error("simulated wire state changed");
}


// --8<-- [start:io]
const copy = Schematic.fromData(readFileSync("beacon.schem"));
copy.setBlock(0, 2, 0, "minecraft:glass");
const editedBytes = bytesFromBase64(copy.toLitematicB64());
writeFileSync("beacon-edited.litematic", editedBytes);
// --8<-- [end:io]

if (editedBytes.length === 0 || copy.blockCount() !== 11) {
  throw new Error("file round trip failed");
}
console.log("Basics JavaScript examples: OK");
